chore(admin): convert manual release creation steps into automated script

timotheeg opened the pull request:
## Problem

Preparing for release is tedious with manual steps to cut branches, extract version, tag, push to staging, copy test procedures, and opening up 2 PRs (release OR to master, backport PR to develop).

While not a huge drain of time, the procedures has been in place for years, and time adds up. An automation script can cut the prep time to just a few seconds.


## Solution

Provide a bash prep scripts engineers can run to start on a release in seconds.

This PR includes no runtime code for production.


Can do a first run with @dcshzj for next release.

seaerchin commented on scripts/release_prep.sh:
```diff
@@ -0,0 +1,122 @@
+
+release_version=$(npm --no-git-tag-version version minor | grep -E '^v\d')
+release_branch=release_${release_version}
+may_force_push=
```
you seem to have forgotten to assign here! (maybe we set this to `false`?)

seaerchin replied:
ahok thanks! don't really write bash so i suck at this AHHA

seaerchin commented on scripts/release_prep.sh:
```diff
@@ -0,0 +1,122 @@
+pr_body_file=.pr_body_${release_version}
+pr_body_file_groupped=.pr_body_${release_version}_groupped
+
+awk "/^#### \[${release_version}\]/{flag=1;next}/####/{flag=0}flag" CHANGELOG.md | sed -E '/^([^-]|[[:space:]]*$)/d' > ${pr_body_file}
```
can we add a comment here? not entirely sure i understand what this is doing HAHA

timotheeg replied:
Can! It's loading the entire section of changes from the current release. If the current release is v0.71.0, it basically grabbing everything that's in between the headings
```
#### v0.71.0
- fix:...
- chore:...
...

#### v0.70.0
```

and it discards lines which only contain whitespaces and empty bullet points.

timotheeg replied:
added a comment: https://github.com/isomerpages/isomercms-backend/pull/1209/commits/9653429ff980cfafcdb72246de5f31c3e2ce81c7

seaerchin commented on scripts/release_prep.sh:
```diff
@@ -0,0 +1,122 @@
+
+# extract changelog to inject into the PR
+pr_body_file=.pr_body_${release_version}
+pr_body_file_groupped=.pr_body_${release_version}_groupped
```
nit: grouped

timotheeg replied:
omg 😱

timotheeg replied:
fixed https://github.com/isomerpages/isomercms-backend/pull/1209/commits/fb45008eec6e5c4e85d1cea1b65854d1fd9f8769

seaerchin commented on scripts/release_prep.sh:
```diff
@@ -0,0 +1,122 @@
+echo "" >> ${pr_body_file_groupped}
+echo "## Tests" >> ${pr_body_file_groupped}
+echo "" >> ${pr_body_file_groupped}
+grep -v -E -- '- [a-z]+\(deps(-dev)?\)' ${pr_body_file} | grep -v -E -- '- (release|backport) v' | grep -v -E -- '- \d+\.\d+\.\d+ ' | while read line_item; do
+  pr_id=$(echo ${line_item} | grep -o -E '\[`#\d+`\]' | grep -o -E '\d+')
+  pr_content=$(gh pr view ${pr_id})
+  tests=$(echo "$pr_content" | awk "/^## Test/{flag=1;next}/^## /{flag=0}flag" | sed -E "s/\[[Xx]\]/[ ]/" | sed -E "s/^(###+) /\1## /")
+  if [[ ${tests} =~ [^[:space:]] ]]; then
+    echo ${line_item} | sed "s/^- /### /" >> ${pr_body_file_groupped}
+    echo "${tests}" >> ${pr_body_file_groupped}
+    echo "" >> ${pr_body_file_groupped}
+  fi
```
could we add a comment here :monkas:

seaerchin reviewed (commented):
not approving yet so we can add comments but functionality wise no blockers.

seaerchin commented:
could you also dup this to FE once comments added!

kishore03109 commented on scripts/release_prep.sh:
```diff
@@ -0,0 +1,122 @@
+grep -v -E -- '- [a-z]+\(deps(-dev)?\)' ${pr_body_file} | grep -v -E -- '- (release|backport) v' | grep -v -E -- '- \d+\.\d+\.\d+ ' | while read line_item; do
+  pr_id=$(echo ${line_item} | grep -o -E '\[`#\d+`\]' | grep -o -E '\d+')
+  pr_content=$(gh pr view ${pr_id})
+  tests=$(echo "$pr_content" | awk "/^## Test/{flag=1;next}/^## /{flag=0}flag" | sed -E "s/\[[Xx]\]/[ ]/" | sed -E "s/^(###+) /\1## /")
```
omg this is gold

timotheeg replied:
Hehe thanks! But you sure though? 

I thought the reason you guys kept the process of "copy tests manually from all PRs into the release PR" for so many years is because you loooooved copy/pasting stuff, no? 😜

I wouldn't want to deprive you of some good old copy/paste "pleasures" 😜

kishore03109 replied:
tim thats enough beat down please :(

kishore03109 reviewed (approved):
this is a life saver, thank you

timotheeg commented:
@dcshzj just ran this in for the current release FE+BE. Everything worked well! ✌️ 💪 

![image](https://github.com/isomerpages/isomercms-backend/assets/935223/1066645e-39a2-4e25-8b29-330fecb6f302)

These are the 4 PRs that were created (in just a few seconds!) for reference:
* https://github.com/isomerpages/isomercms-backend/pull/1213 (BE release PR)
* https://github.com/isomerpages/isomercms-backend/pull/1214 (BE backport PR)
* https://github.com/isomerpages/isomercms-frontend/pull/1845 (FE release PR)
* https://github.com/isomerpages/isomercms-frontend/pull/1846 (FE backport PR)


Some notes and TODOs:
* The script is not creating the draft deployments, it should do that too
* Several PRs did not have a test sections when they probably should have (e.g. [this one](https://github.com/isomerpages/isomercms-frontend/pull/1828))
* Tests for [one PR](https://github.com/isomerpages/isomercms-backend/pull/1199) could not be captured because the test section did not follow the PR template naming convention. section title was `Manual Local Tests`

kishore03109 commented:
@timotheeg 
haha intended! since the tests cannot be performed on staging easily, so not part of deployment process haha

kishore03109 commented on scripts/release_prep.sh:
```diff
@@ -0,0 +1,122 @@
+  exit 1
+fi
+
+git fetch --all --tags
```
@timotheeg bump on this tho!

timotheeg replied:
Hmm, I reckon we shouldn't reset all local tags (in case you have whatever tags specific to your local env), but all tags from origin should indeed overwrite conflicting local tags, yes! (origin is source of truth!)

It's not blocking though, when @dcshzj run the script earlier, there was a conflict for a tag `v0.69.0` he had locally that was different from origin, but the script just moved on anyway. The local tags are not used by the script, the only time things could go wrong if there's already a tag for the one the script try to create. 

For that case, there's a `--recut` option btw.

So hmmm, it's not the purpose of the script to clean your local tags, but I suppose it shouldn't be hard to add that cleanup as well 🤔

timotheeg replied:
done: https://github.com/isomerpages/isomercms-backend/pull/1209/commits/1382944fe43676880fe8a8565597c3be6d7b74e9

dcshzj commented:
Sorry I forgot one thing to point out for @timotheeg, are we able to copy over the Deploy notes section as well? These are steps to perform when doing the release (e.g. creating alarms, updating SSM, etc)

timotheeg commented:
We can! If the convention is well followed by all PRs, we can carry the deploy notes into the release PR!

timotheeg commented:
* Added deploy notes to PR
* Added creation of draft release
* cleaned up the code a bit to be more readable

There are comments that indicate the intent of portions the script, and the variable names should be explanatory enough to indicate what they contain. 

I'm not going to add comments that explains what the bash commands themselves do. That would be much too noisy and not add much value. So if you want info on the bash commands (sed/awk/whatever, feel free to ping me)

There's not much to understand really:
* compute the changelog via `npm version minor`
* each line item contains a pr id that can be extracted
* use the github cli to get PR info for each id
* find content in between sections by convention (e.g. `## Tests` section and `## Deploy Notes` section) as needed and append to local files

timotheeg commented:
Added syncing local tags to origin tags

timotheeg commented:
I added a few more comments which I think should suffice.

I'll merge now, and I'll pair with Chin @seaerchin to run the script for this week's release.
